String interning

Eh2406 opened the pull request:
This builds on the work from #5118. This interns the strings in the part of resolver that gets cloned a lot.

In a test on https://github.com/rust-lang/cargo/issues/4810#issuecomment-357553286
Before we got to 1700000 ticks in ~(63 to 67) sec from #5118
After we got to 1700000 ticks in ~(42 to 45) sec

The interning code itself would be much better with a `leak` function that converts a `String` to a `&'static str`. Something like:
```rust
pub fn leek(s: String) -> &'static str {
    let ptr = s.as_ptr();
    let len = s.len();
    mem::forget(s);
    unsafe {
        let slice = slice::from_raw_parts(ptr, len);
        str::from_utf8(slice).unwrap()
    }
}
```
but "there is no unsafe in Cargo", and I am not the best at unsafe. So I just `to_string` and lived with the extra copy. Is there a better way to hand out references?

I assumed that `InternedString::new` world start appearing in profile result, and that we would want `PackageId`, and `Summary`, Et Al. to store the `InternedString`. That is why I put the interner in a shared folder. So far it is just used in the resolver. It may make sense for a lot more of the Strings to be interned, but with the extra copy... I have not explored it yet.

matklad commented:
Mostly a drive-by comment, but I wonder if we should use strings for features at all... A natural idea is to enumerate all  features in a package, and then use a number as a feature identifier. That way, a *set* of features could be stored as a bitset and fit into a single u64 most of the time....

Eh2406 commented:
@matklad So I see one small problem with that, If I require ```winapi = {version = "0.3", features = ["ntdef"]}``` how do I store the `ntdef` before we have picked a version of `winapi`? Maybe different versions of `winapi` store it as different numbers.

alexcrichton commented on src/cargo/core/interning.rs:
```diff
@@ -0,0 +1,34 @@
+        RwLock::new((Vec::new(), HashMap::new()));
+}
+
+#[derive(Ord, PartialOrd, Eq, PartialEq, Hash, Clone, Copy)]
```
Could the `Ord` and `PartialOrd` references be left out here? I think b/c they're comparing strings rather than contents it may end up being confusing down the road.

Eh2406 commented:
`servo>cargo generate-lockfile -Z no-index-update` Still runs in ~34 sec most of that in many ```Updating git repository `https://github.com/servo/...` ```

I don't really know what I think about having a `Interner` structure. I don't look forward to trying to thread an argument to the `Interner` to every place we want to make an `InternedString` whatever scope we decide should own the `Interner`. I also don't like having to think about if an `InternedString` managed to outlive its backing `Interner`, nor if an `InternedString` is being used with the correct `Interner`. On the other hand, the static/leak strategy... well... it leaks witch is not grate, and it is spooky action at a distance. So I see the advantage in "not that".

The current implementation only uses `InternedString` in the resolver. So the conservative thing to do is to just make the `Interner` an argument to `activate_deps_loop`. But, most of our new `InternedString` come from `InternedString::new(x.package_id().name())` so it would be nice for `PackageId` to already know the `InternedString` of its name, not that it shows up in profiles.

If we make `Interner` a part of `Config` would that help for projects using cargo as a library? Do they create a new `Config` on a regular basis? Can we be sure that they won't use a result from one config with a different config?

I did not understand your suggestion about `scoped thread local`. Sorry.
If you are suggesting using `thread_local!` instead of `lazy_static!`, dose that help with projects using cargo as a library? Do they create a new thread on a regular basis?

So overall I don't know.

alexcrichton commented:
Ok good to know on the resolve time! Sorry I think I wasn't clear before, but on second thought it's probably not worth it to have destruction here and it's probably best to just have this be a global. We can always work with it later and tweak it as necessary.

If it's global though we can for sure never delete it, so want to go ahead and add a `Deref` implementation? That should be safe as the strings basically live for the static lifetime anyway.

Eh2406 commented:
Just to confirm you are leaning toward using the global/leek strategy, and you are asking me to add code that uses unsafe to convert an `InternedString` -> `&'static str`?

Next time I get a chance I will play with some of the options, and put links hear for use to compare.
Specifically, Adding a `Interner` owned by `activate_deps`. It may not be as intrusive as I feare.

alexcrichton commented:
Ah yeah I'm coming around to a global interner which leaks everything you add to it. I do like the idea of `InternedString` though rather than `&'static str` because it makes equality testing an integer comparison rather than a full string comparison.

Eh2406 commented:
Note: that all the performance number should be rethought in lite of #5132 If we are cloning less often than improvements to performance of cloning will matter less.

I assume that if two `&str`s have the same `ptr` and `len` then the contents are never checked. But a dedicated `InternedString` also is fast for !eq tests.

Eh2406 commented:
Pre #5132:
1700000 ticks in ~(63 to 67) sec from #5118 0r 26k ticks/sec
With #5132, still without this pr:
5000000, ticks in ~72 sec, 0r 69k ticks/sec
So that Is a lot better! And the profile is still mostly Cloning activations.
Next to rebase this pr :-)

alexcrichton commented:
Holy cow, nice!

Eh2406 commented:
A nice to you. You are the one that spotted and identified the problem! I didn't notice that the clones I was optimizing where not supposed to be there.

I rebased and clean up this PR. Remarkably, it still pulls its weight!
After this pr, 5000000 ticks in ~52 sec, 0r 96k ticks/sec

Next to try a version with `unsafe` and a separate version with an `Interner` so we can decide how to proceed. :-)

Eh2406 commented:
I have worked on the scoped/`Interner` version. The ergonomics of passing the `Interner` to all the places we need to convert from a `&str` are abysmal. For example `RemainingCandidates.next` neads to compre the links attribute of a `summary` to a hashset of `InternedString`s, but `summary`s out live the `Interner` so they have to return strings, I.E. `RemainingCandidates.next` needs to have a `Interner` argument.

I have threaded it thru to some of the places that need it, ~~and am so bored I just can't finish this experiment.~~ and you can see the result at https://github.com/Eh2406/cargo/commit/0bdbc08ba6e2b3b7731d14256b4fd90014016494. Now that it is done it is not so bad. :-)

alexcrichton commented:
Thanks for trying that out! I think it's clear the lazy_static approach will do fine for now. If you additionally don't see lock contention in the profile then I think it's for sure a fine approach.

Eh2406 commented:
The `unsafe` version is a very small change:
https://github.com/Eh2406/cargo/commit/26bfc7de367860794ea81f468a44ae68cd6deb70

However the fact that this can hand out `&str` at 0 cost means we can use it in the things resolver talk to, so next let us see if that can make a difference. Most of the time is in `clone` so probably not, but let's give it a try. Yes, we can make a difference:
https://github.com/Eh2406/cargo/commit/411355aca37fdd8b361481a350c3cb625beabd1a

Witch brings us to: 5000000 ticks in ~48 sec, 0r 104k ticks/sec
If we go this route, I should try Interning all the things related to `features`, but that was too big for this test.

alexcrichton commented:
Ok sounds good to me! And to be clear it's not that we want no unsafe code in Cargo we just want very little :). Cargo itself has a fair bit of transitive unsafe code

Eh2406 commented:
I tried interning the features:
https://github.com/Eh2406/cargo/commit/6360f25b61eb64cee40abc8f7a71c6304a7646ba
But despite all the changes that involved it was not measurably faster.

If we want the speed of https://github.com/Eh2406/cargo/commit/411355aca37fdd8b361481a350c3cb625beabd1a and dont want the `unsafe` we can add a `interned_name` fealed to `Dependency` so it can hand out both.

One way or the other, I am feeling ready to move onto the next avenue of attack.

alexcrichton commented:
Ah yeah I'm fine incrementally improving the situation here rather than trying to do it all at once. Did you want to put some more changes in this PR and then I can r+?

alexcrichton commented on src/cargo/core/interning.rs:
```diff
@@ -0,0 +1,67 @@
+use std::mem;
+use std::cmp::Ordering;
+
+pub fn leek(s: String) -> &'static str {
```
One thing this may actually want to do is call `s.into_boxed_str()` as that'll call `shink_to_fit` and make sure we're not hanging on to any lingering bytes, but that can happen in a follow-up!

Eh2406 replied:
```
    let boxed = s.into_boxed_str();
    let ptr = boxed.as_ptr();
    let len = boxed.len();
    mem::forget(boxed);
```
Yes?

alexcrichton replied:
indeed!

Eventually this could even use `Box::leak` :)

alexcrichton commented on src/cargo/core/interning.rs:
```diff
@@ -0,0 +1,67 @@
+        cache.insert(s);
+        InternedString { ptr: s.as_ptr(), len: s.len() }
+    }
+    pub fn to_inner(&self) -> &'static str {
```
Could this be listed as an implementation of `Deref` to get all the fun methods on strings?

Eh2406 replied:
Yes, thou they do not hash the same. I can use `Deref` instead of `to_inner` if you prefer.

alexcrichton replied:
Ah yeah I wouldn't worry too much about that

alexcrichton commented:
Looks great! I'd like to get the `Deref` change in before merging but otherwise looks good to go!

alexcrichton commented:
Er actually, there's also a bunch of test failures on Windows at least :(

Eh2406 commented:
I removed the last optimization as I don't know how it was causing the test failures. And added your suggestions. I think let us merge this, and I'll rework the last commit as a separate PR.

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit 98480e8 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 98480e8591d30575daedf738d37aa4488050547a -->

bors commented:
:hourglass: Testing commit 98480e8591d30575daedf738d37aa4488050547a with merge 8fde4e3fec666b219410ae2d4945c110e9b8aa1e...
